### PR TITLE
[feature] Service worker

### PR DESCRIPTION
**Pull request by KonradKlimczak**

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -19,3 +19,17 @@ ReactDOM.render(
   </Provider>,
   document.getElementById('app'),
 );
+
+// Register a service worker
+
+if ('serviceWorker' in navigator) {
```
Comment (BANASZCZYK-GFT-DA) [inline src/index.tsx]: Chujowe

File: src/sw.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+self.addEventListener('fetch', function (event) {
+  // Only fall back for HTML documents.
+  var request = event.request;
+  // && request.headers.get('accept').includes('text/html')
+  if (request.method === 'GET') {
+    // `fetch()` will use the cache when possible, to this examples
+    // depends on cache-busting URL parameter to avoid the cache.
+    event.respondWith(
+      fetch(request).catch(function (error) {
+        // `fetch()` throws an exception when the server is unreachable but not
```
Comment (BANASZCZYK-GFT-DA) [inline src/sw.js]: Chujowe

File: webpack.config.js [hunk 3]
```diff
@@ -57,4 +60,27 @@ module.exports = {
       chunkFilename: '[id].css'
     })
   ]
-};
+}, {
+  entry: './src/sw.js',
+  output: {
+    globalObject: 'this',
+    path: path.resolve(__dirname, 'dist'),
```
Comment (BANASZCZYK-GFT-DA) [inline webpack.config.js]: Ja pierdolę

File: webpack.config.js [hunk 4]
```diff
@@ -57,4 +60,27 @@ module.exports = {
...
+  },
+  resolve: {
+    extensions: ['.js']
+  },
+  module: {
+    rules: [{
+      test: /\.js$/,
+      exclude: /node_modules/,
+      use: {
+        loader: 'babel-loader'
```
Comment (BANASZCZYK-GFT-DA) [inline webpack.config.js]: Chujowe

File: src/sw.js [hunk 5]
```diff
@@ -1,39 +1,42 @@
...
-  var offlineRequest = new Request('offline.html');
-  event.waitUntil(
-    fetch(offlineRequest).then(function (response) {
-      return caches.open('offline').then(function (cache) {
-        console.log('[oninstall] Cached offline page', response.url);
-        return cache.put(offlineRequest, response);
-      });
-    })
-  );
+const CACHE = 'cache-and-update';
```
Comment (enril) [inline src/sw.js]: Rename file. I know, that sw equals serviceWorker, but we should use full sentences not only first letters or magic variables :P

File: src/index.tsx [hunk 6]
```diff
@@ -19,3 +19,17 @@ ReactDOM.render(
...
+  (window || {} as any).addEventListener('load', function () {
+    navigator.serviceWorker.register('/sw.js').then(function (registration) {
+      // Registration was successful
```
Comment (enril) [inline src/index.tsx]: You have the same information twice. Just console log. :)

File: src/index.tsx [hunk 7]
```diff
@@ -19,3 +19,17 @@ ReactDOM.render(
...
+      console.log('ServiceWorker registration successful with scope: ', registration.scope);
+    }, function (err) {
+      // registration failed :(
```
Comment (enril) [inline src/index.tsx]: You have the same information twice. Just console log. :)

File: src/index.tsx [hunk 8]
```diff
@@ -19,3 +19,17 @@ ReactDOM.render(
   </Provider>,
   document.getElementById('app'),
 );
+
+// Register a service worker
+
+if ('serviceWorker' in navigator) {
+  (window || {} as any).addEventListener('load', function () {
+    navigator.serviceWorker.register('/sw.js').then(function (registration) {
```
Comment (enril) [inline src/index.tsx]: This is the format proposal, but for me, we should put enter before then. This code will look better.

File: src/sw.js (hunk 2 above)
Comment (enril) [inline src/sw.js]: Ok Just removed uncoment code or write the description, why we have to have this comment.

**Pull request merged**